### PR TITLE
Clean up documentation

### PR DESCRIPTION
**Pull request by nomis52**
This switches to markdown format so it's easier for other people to edit.

### CODE REVIEW

**Comment by peternewman** [comment]
Can you resync this please.

**Comment by nomis52** [comment]
Done

**Comment by nomis52** [comment]
Lint will come right once #164 is in

File: README.md [hunk 1]
```diff
@@ -30,6 +21,26 @@ The Ja Rule codebase is licensed under the
 The unit-testing code & mocks are licenced under the
 [GPL](http://www.gnu.org/licenses/gpl.html).
 
+The hardware designs and the documentation is licensed under the
+[Creative Commons BY-SA](https://creativecommons.org/licenses/by-sa/3.0/us/).
+
+## Directory Layout
+
+```
+├── Bootloader  # The DFU bootloader
```
Comment (peternewman) [inline README.md]: Add the boardcfg folder.

File: README.md [hunk 2]
```diff
@@ -30,6 +21,26 @@ The Ja Rule codebase is licensed under the
...
+│       ├── Bootloader.X  # Bootloader MPLAB X project
+│       ├── src  # Bootloader source code
+├── common  # Common code shared between the bootloader and application.
+├── firmware  # The main DMX/RDM application
+│   ├── ja-rule.X
+│   └── src
+├── linker       # linker scripts for the bootloader & application
+├── tests        # Unit tests
+├── tools        # tools to upgrade the firmware on the device.
+└── user_manual  # The user manual documentation
```
Comment (peternewman) [inline README.md]: I'd say documentation is redundant here, unless it's the instructions for how to write or use the user manual; the user manual manual.

File: tests/Doxyfile [hunk 3]
```diff
@@ -657,6 +657,17 @@ INPUT                  = ../Bootloader/firmware/src \
                          ../common \
                          ../firmware/src \
                          ../firmware/src/system_config/ethernet_sk2/system_settings.h \
+                         doxygen/main.md \
```
Comment (peternewman) [inline tests/Doxyfile]: We have to explicitly add md files do we? Can we not glob?

File: tests/doxygen/bootloader.md [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+. This makes it easier for end users to upgrade their devices.
+
+The bootloader uses the <a
+[Device Firmware Upgrade (DFU)]
+(http://www.usb.org/developers/docs/devclass_docs/DFU_1.1.pdf)
+protocol. This is a standard protocol for upgrading firmware over USB.
+
+@section bootloader_procedure Update Procedure
+
+Generally the update procedure follows:
```
Comment (peternewman) [inline tests/doxygen/bootloader.md]: This feels like it's missing a word or something. procedure involves the following steps perhaps?

File: tests/doxygen/main.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ - Mark time
+ - Response delay
+
+## DMX512 / RDM Receiver {#main-features-responder}
+
+- The ability to simulate many different types of @ref rdm-models.
+ Together, they support all PIDs from the E1.20, E1.37-1 & E1.37-2
+ standards. This allows RDM controllers to be tested without having access
+ to a large number of RDM devices. The code can act as:
+ - An LED Driver
```
Comment (peternewman) [inline tests/doxygen/main.md]: Can we make the capitalisation consistent, either drop driver, moving light, network device, or promote the others.

File: tests/doxygen/main.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+ - A Moving Light
+ - A proxy, which will ACK_TIMER all requests to the child devices.
+ - A sensor only device (No DMX footprint)
+ - A dimmer, with sub-devices including all PIDs from E1.37-1
+ - An Network Device, including all PIDs from E1.37-2.
```
Comment (peternewman) [inline tests/doxygen/main.md]: SPaG A network

File: tests/doxygen/main.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+
+# Supported Platforms {#main-platforms}
+
+This code is targeted at the PIC32 series MCUs. It uses the parts of the
+[Harmony](http://www.microchip.com/pagehandler/en_us/devtools/mplabharmony/home.html)
+framework, so it won't work on other platforms without some porting effort.
+
+# Terminology {#main-terminology}
+
+<dl>
```
Comment (peternewman) [inline tests/doxygen/main.md]: No markdown equivalent for this then?

File: tests/doxygen/main.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+<dt>Device</dt>
+<dd>The embedded system running this code.</dd>
+<dt>Host</dt>
+<dd>The system that typically runs a full OS, and is the USB / SPI bus
```
Comment (peternewman) [inline tests/doxygen/main.md]: Is it ever SPI bus master?

File: tests/doxygen/main.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+    master.</dd>
+<dt>%Message</dt>
```
Comment (peternewman) [inline tests/doxygen/main.md]: Is % a link or something?

File: tests/doxygen/message-format.md [hunk 10]
```diff
@@ -0,0 +1,697 @@
...
+@param Token A token for the command. The same token will be returned in the
+command response. Typically the host will increment the token with each
+request.
+@param Command The @ref Command identifier.
+@param Length The length of the data included in the command. The valid
+range is 0 - 579 bytes.
+@param Payload The payload data associated with the command. See each
+command type below for the specific format of the payload data.
+@param EOM The end of message identifier: @ref END_OF_MESSAGE_ID
+@param Padding Extra padding, the Host should set this to 0. Padding should
```
Comment (peternewman) [inline tests/doxygen/message-format.md]: Clarify, the host should not use this, or should pad with zeros, to what size?

File: tests/doxygen/message-format.md [hunk 11]
```diff
@@ -0,0 +1,697 @@
...
+request.
+@param Command The @ref Command identifier.
+@param Return_Code The @ref ReturnCode of the command.
+@param Status The status bitfield.
+@param Length The length of the data included in the command. The valid
+range is 0 - 579 bytes.
+@param Payload The payload data associated with the command. See each
+command type below for the specific format of the payload data.
+@param EOM The end of message identifier: @ref END_OF_MESSAGE_ID
+@param Padding Extra padding, the Host should set this to 0. Padding should
```
Comment (peternewman) [inline tests/doxygen/message-format.md]: If this is a response, it must be the device setting it surely?

File: tests/doxygen/message-format.md [hunk 12]
```diff
@@ -0,0 +1,697 @@
...
+ /                                                               /
+ \                        Payload (if any)                       \
+ /                                                               /
+ +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ |      EOM      |               Padding (optional)              \
+ +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+</pre>
+
+@param SOM The start of message identifier: @ref START_OF_MESSAGE_ID
+@param Token A token for the command. The same token will be returned in the
```
Comment (peternewman) [inline tests/doxygen/message-format.md]: Correct this.

File: tests/doxygen/message-format.md [hunk 13]
```diff
@@ -0,0 +1,697 @@
...
+command response. Typically the host will increment the token with each
+request.
+@param Command The @ref Command identifier.
+@param Return_Code The @ref ReturnCode of the command.
+@param Status The status bitfield.
+@param Length The length of the data included in the command. The valid
```
Comment (peternewman) [inline tests/doxygen/message-format.md]: Correct

File: tests/doxygen/message-format.md [hunk 14]
```diff
@@ -0,0 +1,697 @@
...
+range is 0 - 579 bytes.
+@param Payload The payload data associated with the command. See each
```
Comment (peternewman) [inline tests/doxygen/message-format.md]: Correct

File: tests/doxygen/message-format.md [hunk 15]
```diff
@@ -0,0 +1,697 @@
...
+@param Timeout The time to wait for a RDM response after a broadcast command,
+in 10ths of a millisecond. See Transceiver_SetRDMBroadcastTimeout().
+
+### Response Payload {#message-commands-setbcasttimeout-res}
+
+The response contains no data.
+
+@returns @ref RC_OK or @ref RC_BAD_PARAM if the value was out of range.
+
+## Get RDM Response Timerout {#message-commands-getresponsetimeout}
```
Comment (peternewman) [inline tests/doxygen/message-format.md]: SPaG Timeout

File: tests/doxygen/message-format.md [hunk 16]
```diff
@@ -0,0 +1,697 @@
...
+
+### Response Payload {#message-commands-setresponderdelay-res}
+
+The response contains no data.
+
+@returns @ref RC_OK or @ref RC_BAD_PARAM if the value was out of range.
+
+## Get RDM Responder Jitter {#message-commands-getresponderjitter}
+
+Get the jitter to use when sending RDM responses.
```
Comment (peternewman) [inline tests/doxygen/message-format.md]: Is this not the maximum jitter time?

File: tests/doxygen/message-format.md [hunk 17]
```diff
@@ -0,0 +1,697 @@
...
+@param RDM_DUB_Response The raw response, if any was received.
+@returns
+- @ref RC_OK if the frame was sent correctly and data was received.
+- @ref RC_BUFFER_FULL if the transmit buffer is full.
+- @ref RC_TX_ERROR if a transmit error occurred.
+- @ref RC_RDM_TIMEOUT if no response was received.
+
+## Transmit Broadcast RDM Get / Set {#message-commands-txrdmbroadcast}
+
+Sends a broadcast RDM Get / Set command. If the Broadcast Listen Delay is 0,
```
Comment (peternewman) [inline tests/doxygen/message-format.md]: Shouldn't this be non-zero?

File: tests/doxygen/message-format.md [hunk 18]
```diff
@@ -0,0 +1,697 @@
...
+the start of the break, in 10ths of a microsecond. Undefined unless
+RC_RDM_BCAST_RESPONSE was returned.
+@param Break_End the time from the end of the transmitted RDM frame to
+the end of the break / start of the mark, in 10ths of a microsecond.
+Undefined unless RC_RDM_BCAST_RESPONSE was returned.
+@param Mark_End the time from the end of the transmitted RDM frame to
+the start of the mark, in 10ths of a microsecond. Undefined unless
+RC_RDM_BCAST_RESPONSE was returned.
+@param RDM_Response The RDM response, if any was received.
+@returns
```
Comment (peternewman) [inline tests/doxygen/message-format.md]: What happens if I send a unicast command using this command?

File: tests/doxygen/mplab-setup.md [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+(Simon) have concerns about code quality.
+
+MPLAB X comes with a [Harmony Configurator Plugin (MHC)](
+https://microchip.wikidot.com/harmony:mhc-installation)
+which auto-generates the boiler plate code from a Harmony configuration.
+
+When used with Harmony, the design of the MPLAB X build / configuration
+system makes it difficult to move a project's directory to a new location on
+the filesystem. In fact, by default, a project's configurations.xml file
+uses relative paths!
```
Comment (peternewman) [inline tests/doxygen/mplab-setup.md]: Don't you mean absolute?

File: tests/doxygen/rdm_models.md [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+are 'hidden' in a sense.
+
+Using the OLA CLI, to get a list of available models:
+~~~~~
+$ ola_rdm_get -u 1 --uid 7a70:fffffe00 model_id_list
+~~~~~
+
+To change the current model:
+~~~~~
+$ ola_rdm_set -u 1 --uid 7a70:fffffe00 model_id 258
```
Comment (peternewman) [inline tests/doxygen/rdm_models.md]: Is it worth giving a valid example model ID?

**Comment by peternewman** [comment]
Some comments, but generally okay.

File: README.md (hunk 1 above)
Comment (nomis52) [inline README.md]: Done

File: README.md (hunk 2 above)
Comment (nomis52) [inline README.md]: Removed

File: tests/Doxyfile (hunk 3 above)
Comment (nomis52) [inline tests/Doxyfile]: The order the files appear in controls the section ordering in the output. I haven't worried about ordering for now, once I write an intro I'll clean it up.

Comment (nomis52) [inline tests/Doxyfile]: I added a comment

File: tests/doxygen/bootloader.md (hunk 4 above)
Comment (nomis52) [inline tests/doxygen/bootloader.md]: Fixed

File: tests/doxygen/main.md (hunk 5 above)
Comment (nomis52) [inline tests/doxygen/main.md]: Done

File: tests/doxygen/main.md (hunk 6 above)
Comment (nomis52) [inline tests/doxygen/main.md]: Done

File: tests/doxygen/main.md (hunk 7 above)
Comment (nomis52) [inline tests/doxygen/main.md]: Nope

File: tests/doxygen/main.md (hunk 8 above)
Comment (nomis52) [inline tests/doxygen/main.md]: Not yet. That code is coming.

File: tests/doxygen/main.md (hunk 9 above)
Comment (nomis52) [inline tests/doxygen/main.md]: Yes

File: tests/doxygen/message-format.md (hunk 10 above)
Comment (nomis52) [inline tests/doxygen/message-format.md]: Done

File: tests/doxygen/message-format.md (hunk 12 above)
Comment (nomis52) [inline tests/doxygen/message-format.md]: Done

File: tests/doxygen/message-format.md (hunk 13 above)
Comment (nomis52) [inline tests/doxygen/message-format.md]: ??

File: tests/doxygen/message-format.md (hunk 14 above)
Comment (nomis52) [inline tests/doxygen/message-format.md]: ??

File: tests/doxygen/message-format.md (hunk 11 above)
Comment (nomis52) [inline tests/doxygen/message-format.md]: Fixed

File: tests/doxygen/message-format.md (hunk 15 above)
Comment (nomis52) [inline tests/doxygen/message-format.md]: Fixed

File: tests/doxygen/message-format.md (hunk 16 above)
Comment (nomis52) [inline tests/doxygen/message-format.md]: fixed

File: tests/doxygen/message-format.md (hunk 17 above)
Comment (nomis52) [inline tests/doxygen/message-format.md]: Fixed

File: tests/doxygen/message-format.md (hunk 18 above)
Comment (nomis52) [inline tests/doxygen/message-format.md]: It depends on what the broadcast delay was. Either the response will be ignored or it'll cause an error to be returned

File: tests/doxygen/mplab-setup.md (hunk 19 above)
Comment (nomis52) [inline tests/doxygen/mplab-setup.md]: No, it defaults to relative, we try to force it to absolute.

File: tests/doxygen/rdm_models.md (hunk 20 above)
Comment (nomis52) [inline tests/doxygen/rdm_models.md]: I'm going to let michael clean this up. This part will go into the user doc anyway.

File: tests/doxygen/main.md (hunk 8 above)
Comment (peternewman) [inline tests/doxygen/main.md]: Ah, but it will come in future?

File: tests/Doxyfile (hunk 3 above)
Comment (peternewman) [inline tests/Doxyfile]: Ah okay.

File: tests/doxygen/main.md (hunk 8 above)
Comment (nomis52) [inline tests/doxygen/main.md]: That's the plan

File: tests/Doxyfile (hunk 3 above)
Comment (peternewman) [inline tests/Doxyfile]: Out of interest, what happens if you add and then add again via a glob?

File: tests/doxygen/message-format.md (hunk 13 above)
Comment (peternewman) [inline tests/doxygen/message-format.md]: Its a response not a command, was I think what I meant.

File: tests/doxygen/message-format.md (hunk 14 above)
Comment (peternewman) [inline tests/doxygen/message-format.md]: Again response.

File: tests/doxygen/mplab-setup.md (hunk 19 above)
Comment (peternewman) [inline tests/doxygen/mplab-setup.md]: Ah, I'd have thought relative would be okay (apart from the Harmony stuff).

File: tests/doxygen/rdm_models.md (hunk 20 above)
Comment (peternewman) [inline tests/doxygen/rdm_models.md]: Okay.

**Comment by peternewman** [comment]
Just a few outstanding comments.

File: tests/doxygen/message-format.md [hunk 21]
```diff
@@ -0,0 +1,697 @@
...
+command response. Typically the host will increment the token with each
+request.
+@param Command The @ref Command identifier.
+@param Length The length of the data included in the command. The valid
+range is 0 - 579 bytes.
+@param Payload The payload data associated with the command. See each
+command type below for the specific format of the payload data.
+@param EOM The end of message identifier: @ref END_OF_MESSAGE_ID
+@param Padding Extra padding. If present this should be filled with 0s.
+Padding can be added until the total message does not exceed
```
Comment (peternewman) [inline tests/doxygen/message-format.md]: SPaG as long as the total message does not exceed (or similar)

File: tests/doxygen/message-format.md [hunk 22]
```diff
@@ -0,0 +1,697 @@
...
+@param Command The @ref Command identifier.
+@param Return_Code The @ref ReturnCode of the command.
+@param Status The status bitfield.
+@param Length The length of the data included in the command. The valid
+range is 0 - 579 bytes.
+@param Payload The payload data associated with the command. See each
+command type below for the specific format of the payload data.
+@param EOM The end of message identifier: @ref END_OF_MESSAGE_ID
+@param Padding Extra padding. If present this should be filled with 0s.
+Padding can be added until the total message does not exceed
```
Comment (peternewman) [inline tests/doxygen/message-format.md]: Again.

File: tests/doxygen/main.md [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+ to a large number of RDM devices. The code can act as:
+ - An LED driver
+ - A moving light
+ - A proxy, which will ACK_TIMER all requests to the child devices.
+ - A sensor only device (No DMX footprint)
+ - A dimmer, with sub-devices including all PIDs from E1.37-1
+ - A network device, including all PIDs from E1.37-2.
+- Configurable RDM response delay, with an option to introduce jitter
+- Identify & Mute status indicators.
+- RGB Pixel control using SPI (LPD8806 only for now).
```
Comment (peternewman) [inline tests/doxygen/main.md]: SPaG pixel

**Comment by peternewman** [comment]
A few bits still to fix in here @nomis52 .

File: tests/doxygen/main.md (hunk 23 above)
Comment (nomis52) [inline tests/doxygen/main.md]: Already fixed

File: tests/doxygen/message-format.md (hunk 22 above)
Comment (nomis52) [inline tests/doxygen/message-format.md]: Fixed

File: tests/doxygen/message-format.md (hunk 21 above)
Comment (nomis52) [inline tests/doxygen/message-format.md]: Fixed

**Comment by nomis52** [comment]
Sorry I fixed them all but didn't mark them as done.

**Pull request merged**